### ISSUE TITLE
[PyTorch] random_ on floating type with negative values is broken on both CPU and GPU

### ISSUE THREAD

**Issue by SsnL** [pytorch/pytorch#6338]
Reproduced on master and also broken on 0.3.1.

```python
>>> torch.randn(3,3).random_(-1,10).max().item()
9.0
>>> torch.randn(3,3).random_(-1,10).max().item()
1.8446744073709552e+19
>>> torch.randn(3,3).cuda().random_(-1,10).max().item()
4294967296.0
>>> torch.randn(3,3).cuda().random_(-1,10).max().item()
8.0
>>> torch.randn(3,3).cuda().random_(-1,10).max().item()
4294967296.0
>>> torch.randn(3,3).cuda().random_(-1,10).max().item()
4294967296.0
>>> torch.randn(3,3).cuda().random_(-1,10).max().item()
9.0
>>> torch.randn(3,3).cuda().random_(-1,10).max().item()
9.0
>>> torch.randn(3,3).cuda().random_(-1,10).max().item()
8.0
>>> torch.randn(3,3).cuda().random_(-1,10).max().item()
8.0
>>> torch.randn(3,3).cuda().random_(-1,10).max().item()
8.0
>>> torch.randn(3,3).cuda().random_(-1,10).max().item()
4294967296.0
>>> torch.randn(3,3).cuda().random_(-1,10).max().item()
4294967296.0
```

Reported in #6136

**Comment by albanD** [comment]
I guess the problem here is that the result of `THRandom_random` [here](https://github.com/pytorch/pytorch/blob/81676d8554fe6ae37a1b5e9c637caef8ad2d5f30/aten/src/TH/generic/THTensorRandom.cpp#L40) is unsigned, and we add `min` before casting it into the `real` type.
@SsnL is that the actual issue? Let me know if I should send a PR for that.

**Comment by SsnL** [comment]
@albanD I've verified that this is indeed the problem (at least for CPU). We should cast it into `int64_t` before adding. Changing the line to `static_cast<real>(static_cast<int64_t>(THRandom_random64(_generator) % range) + min)` works. It'd be great if you can submit a PR. :)

**Comment by albanD** [comment]
Do we want to cast it to `int64_t` ? Or `accreal`?

**Comment by albanD** [comment]
Working on this. Will submit the PR tomorrow.

**Comment by SsnL** [comment]
fixed via #6463
